### DISCUSSION TITLE
Request for script to produce MODE related plots

### DISCUSSION THREAD

**Discussion by gbodeImoleayo** [dtcenter/METplus#1413] [Plot Generation] [answered]
Dear All,
Compliments of the very new year and best wishes!

I trust my email finds you well and safe.

I am working on object-based forecast verification using the MODE tool in MET. I have been able to generate the outputs from MODE but would like to present the forecast and observed object plots in a different way.  For example, I want a panel plot showing (i) the raw observed and forecast values on the first row, (ii) all the observed and forecast objects on the second row and (iii) only the matched the observed objects with outline of the matched forecast objects on the third row. I made attempt to produce this type of figure from the nc files on python but the matched forecast outline becomes messy at times when it really matters (see attachment).

Please I will be grateful if you can share a sample script that produces similar plot.

Thank you.

Best regards,
Imoleayo  
![image](https://user-images.githubusercontent.com/47651682/152498313-5b30a979-90ff-48e4-8fce-e0dcff0b8564.png)

![image](https://user-images.githubusercontent.com/47651682/152498408-cb6ad011-360f-43fa-a069-21fe9b7a1b22.png)

**Comment by JohnHalleyGotway** [comment, marked as answer]
@gbodeImoleayo thanks for moving this topic over to GitHub discussions. Looks like you've already put together some nice plots for the output of MODE. I'm glad to see that you've found the MODE NetCDF output files to be useful!

I did make somewhat similar plots of the MODE NetCDF output files myself for a project several years ago. However, they were in NCL and wouldn't provide much help here.

You're actually very close and just need know how to subset the (lat, lon) points when plotting them. First, let me point you to this [table](https://met.readthedocs.io/en/latest/Users_Guide/mode.html#id8) which lists out the names of the variables in the MODE NetCDF output files.

You want to plot "**only the matched the observed objects with outline of the matched forecast objects on the third row**".

In that NetCDF file, the fcst_simp_bdy_lat/lon and obs_simp_bdy_lat/lon variables contain the (lat, lon) locations of the outline of the simple objects. The fcst/obs_simp_hull_lat/lon and fcst/obs_clus_hull_lat/lon variables define the convex hulls (i.e. rubber bands around objects in the MODE plots), but you can ignore these because you want the object outlines instead.

To plot the "**outline of the matched forecast objects**", you'll need to read:
 - fcst_simp_bdy_lat to get one long array of latitudes.
 - fcst_simp_bdy_lon to get one long array of longitudes.
 - fcst_simp_bdy_start to get starting index into the lat/lon arrays
 - fcst_simp_bdy_npts to get the number of points for each object.

Right now, I suspect you're reading fcst_simp_bdy_lat and fcst_simp_bdy_lon and plotting them all in order. Instead, you'll need to plot them separately for each object. By way of example, here's some output from a sample MODE NetCDF file:

```
dimensions:
	fcst_simp = 8 ;
	fcst_simp_bdy = 162 ;
...
variables:
	int fcst_simp_bdy_start(fcst_simp) ;
	int fcst_simp_bdy_npts(fcst_simp) ;
	float fcst_simp_bdy_lat(fcst_simp_bdy) ;
	float fcst_simp_bdy_lon(fcst_simp_bdy) ;
...
data:
 fcst_simp_bdy_start = 0, 19, 44, 75, 114, 119, 128, 157 ;
 fcst_simp_bdy_npts = 19, 25, 31, 39, 5, 9, 29, 5 ;
```

So this example has 8 simple forecast objects (fcst_simp = 8). There are total of 162 lat/lon points provided to describe their outlines (fcst_simp_bdy = 162).
1. For forecast simple object 1, start with fcst_simp_bdy_lat/lon index 0 and plot 19 of them.  
2. For forecast simple object 2, start with fcst_simp_bdy_lat/lon index 19 and plot 25 of them.
3. For forecast simple object 3, start with fcst_simp_bdy_lat/lon index 44 and plot 31 of them.
4. And so on for the remaining objects.

This logic will plot outlines for ALL forecast simple objects. But you'd like to skip over the ones that are unmatched. I can think of 2 choices here.
1. You could parse the _obj.txt file and parse in the OBJECT_ID and OBJECT_CAT columns to determine if an object is matched or unmatched.
  - For example, OBJECT_ID = F001, OBJECT_CAT = CF004 indicates that forecast simple object number 1 is MATCHED and are part of cluster forecast object number 4.
  - Also, OBJECT_ID = F002, OBJECT_CAT = CF000 indicates that forecast simple object number 2 is UNMATCHED since is not part of a cluster forecast object.
2. Alternatively, you could do this entirely within the NetCDF file by reading the fcst_obj_id and fcst_clus_id 2D fields. For simple object number 2, just find one (x, y) value in the fcst_obj_id field that equals 2. Then check the fcst_clus_id value for that (x, y). If it's >0, it's matched. If it's set to -1, then it's unmatched.

With either of these methods, you should be able to skip over the unmatched forecast objects.

**Reply by gbodeImoleayo** [reply]
@JohnHalleyGotway many thanks for the very detailed response to my question.

**Reply by JohnHalleyGotway** [reply]
Great! Thanks for confirming. I'll go ahead and "Lock conversation" on this Discussion. We've found in the past that we lose track of new posts to old discussions that have already been marked as being answered. We prevent that by locking the conversation and encouraging users to create new discussions for new questions.

Thanks - John

**Comment by JohnHalleyGotway** [comment]
@gbodeImoleayo, sorry, looks like I locked this discussion too quickly. Here's the followup email you sent me directly:

----------------------

Regarding the plot of the matched outline on the third raw, I used option 2 "
Alternatively, you could do this entirely within the NetCDF file if you'd like by reading the fcst_obj_id and fcst_clus_id 2D fields. For simple object number 2, just find one (x, y) value in the fcst_obj_id field that equals 2. Then check the fcst_clus_id value for that (x, y). If it's >0, it's matched. If it's set to -1, then it's unmatched.
".
But, I used fcst_obj_id and obs_clus_id instead (see the plot below) because some unmatched objects appears when I use fcst_obj_id and fcst_clus_id. I hope this is correct?

---------------------

And here's my response:

I suspect you're doing this correctly, but do want to make it as clear as possible. Please take a look at these plot_data_plane outputs:
<img width="1001" alt="Screen Shot 2022-02-08 at 9 48 06 AM" src="https://user-images.githubusercontent.com/21087144/153035015-21cc1d18-8d06-4216-a234-20b4cf11a685.png">

These 4 images show the fcst_obj_id (upper left), fcst_clus_id (lower left), obs_obj_id (upper right), and obs_clus_id (lower right). I plotted them all using the same colorscale from -1 to 10. And -1 is plotted as white. Notice in the object id images (top row) that all object id values are > 0. However in the cluster id images (bottom row), some objects are plotted in white, meaning -1. Put simply, the objects with a value of -1 in the cluster field are UNMATCHED.

So for example, for forecast object id number 5, find an (x,y) value in the fcst_obj_id field with a value of 5. Check the value for that (x, y) the fcst_clus_id field. If it's -1, then forecast object id number 5 is UNMATCHED. If its >0, then it is MATCHED.

**Reply by gbodeImoleayo** [reply]
@JohnHalleyGotway 
Thanks for your response and further suggestions.
This is quite helpful.
